feat(core): Add Support for Additional Body Properties in OAuth2 API Client Credentials Flow

DMA902 opened the pull request:
## Summary

Add support for sending additional body properties when using OAuth2 client credentials grant type with body authentication.

This enhancement allows users to include custom properties in the token request body when authenticating with OAuth2 services that require additional parameters beyond the standard client_id and client_secret.

This PR merely exposes it for use in the Generic OAuth2 API Credentials Configuration UI. The backend work to support additional body properties in the token request were done in previous MRs by other contributors:
- https://github.com/n8n-io/n8n/pull/15003
- https://github.com/n8n-io/n8n/pull/15335

This resolves a community pain point that was first expressed back in 2023: https://community.n8n.io/t/oauth2-options-and-troubleshooting/25374.

There are a number of APIs that use OAuth 2.0 Client Credentials Flow which require additional properties in body during the token request (such as audience). Here are a few examples:
- [Pax8](https://devx.pax8.com/reference/createaccesstoken)
- [SPS Commerce](https://developercenter.spscommerce.com/#/docs/new-authentication-docs/machine2machine-applications)
- [Ory](https://www.ory.sh/docs/hydra/guides/audiences)

### Changes
- Add `sendAdditionalBodyProperties` boolean field to OAuth2Api credentials
- Add `additionalBodyProperties` JSON field for custom body parameters
- Update OAuth2 client initialization to include additional body properties when configured
- Display additional body properties field only when grant type is 'clientCredentials', authentication is 'body', and sendAdditionalBodyProperties is enabled

###  Testing
- Verified additional body properties are correctly parsed and included in token requests
- Confirmed field visibility logic works as expected with display options
- Tested with various JSON configurations for additional properties

![Screenshot 2025-06-20 at 23 44 18](https://github.com/user-attachments/assets/7b0d5801-aaf4-4e4c-a642-8f514934120a)

![Screenshot 2025-06-20 at 23 31 22](https://github.com/user-attachments/assets/c85b76d0-edc3-4458-95f6-90045e15cd06)

## Related Linear tickets, Github issues, and Community forum posts

<!--
Include links to **Linear ticket** or Github issue or Community forum post.
Important in order to close *automatically* and provide context to reviewers.
https://linear.app/n8n/issue/
-->
<!-- Use "closes #<issue-number>", "fixes #<issue-number>", or "resolves #<issue-number>" to automatically close issues when the PR is merged. -->


## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [ ] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

DMA902 committed:
Allow Additional Body Properties for Generic OAuth2 API

DMA902 committed:
Merge remote-tracking branch 'upstream/master'

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/n8n-io/n8n?pullRequest=16573) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/n8n-io/n8n?pullRequest=16573) before we can accept your contribution.<br/><hr/>**David Ma** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/n8n-io/n8n?pullRequest=16573) it.</sub>

Joffcom commented:
Hey @DMA902,

Thanks for the PR, We have created "GHC-2680" as the internal reference to get this reviewed.

One of us will be in touch if there are any changes needed, in most cases this is normally within a couple of weeks but it depends on the current workload of the team.

17jmumford commented:
Very excited to see this get in! Will resolve a significant pain point for us.

DMA902 commented:
> Any updates on getting this PR merged @Joffcom?

> Very excited to see this get in! Will resolve a significant pain point for us.

I'm so glad to hear your support of this PR! Glad I'm not the only one facing this challenge 😄

@Joffcom, would love to have the n8n team take a look at this PR when they get a chance. 

I think it'll be a small but mighty PR that will unblock a lot of people from utilizing third party APIs that require additional body properties as part of their OAuth token request (there seems to be quite a few!). It'll make the OAuth 2.0 client credentials flow much more robust in n8n I believe. 

For us personally, it's a bit of a show stopper for utilizing n8n to communicate with these APIs. We'd have to essentially create a custom OAuth workflow to make it work, which seems overkill. 

Happy to make any changes the team needs to get this PR merged!

DMA902 committed:
Add displayOptions for sendAdditionalBodyProperties

DMA902 committed:
Add fallbackValue for additionalBodyProperties

DMA902 commented on packages/nodes-base/credentials/OAuth2Api.credentials.ts:
```diff
@@ -107,6 +107,28 @@ export class OAuth2Api implements ICredentialType {
+		{
+			displayName: 'Send Additional Body Properties',
+			name: 'sendAdditionalBodyProperties',
+			type: 'boolean',
```
Good call! Just added that in and tested.

**`Send Additional Body Properties` does not appear when `Client Credentials` is NOT selected (`Authorization Code` was selected here)** 
<img width="1193" height="743" alt="Screenshot 2025-07-31 at 21 21 52" src="https://github.com/user-attachments/assets/b150991e-a660-45bf-a15a-731d17d89a02" />


**`Send Additional Body Properties` does not appear when `Client Credentials` and `Header` Authentication is selected:**
<img width="1196" height="748" alt="Screenshot 2025-07-31 at 21 08 02" src="https://github.com/user-attachments/assets/d15fcef4-4f4d-4500-b204-a62acc764ede" />


**`Send Additional Body Properties` appears when both `Client Credentials` and `Body` Authentication is selected:**
<img width="1195" height="748" alt="Screenshot 2025-07-31 at 21 08 10" src="https://github.com/user-attachments/assets/99ae0376-c6c8-4843-908e-4cb6fab99063" />

DMA902 commented:
@elsmr Thank you for taking the time to review this PR!

I've addressed your comments in two new commits. Please let me know if there's anything else that should be addressed.

thelastbrittain commented:
Looks great! @Joffcom, anything else before this is ready to be pushed out?

elsmr committed:
Merge branch 'master' into DMA902/master

elsmr committed:
Extract function to reduce complexity
